torchrl/data/tensor_specs.py

```diff
@@ -1014,61 +1027,77 @@ class MultiDiscreteTensorSpec(DiscreteTensorSpec):
+                        device=self.device,
+                        dtype=self.dtype,
+                    )
+                )
+        return torch.cat(x)
+
     def rand(self, shape: Optional[torch.Size] = None) -> torch.Tensor:
         if shape is None:
             shape = torch.Size([])
-        x = torch.cat(
-            [
-                torch.randint(
-                    0,
-                    space.n,
-                    torch.Size([1, *shape]),
-                    device=self.device,
-                    dtype=self.dtype,
-                )
-                for space in self.space
-            ]
-        ).squeeze()
-        _size = [self._size] if self._size > 1 else []
-        return x.T.reshape([*shape, *_size])
 
-    def _split(self, val: torch.Tensor):
-        return [val] if self._size < 2 else val.split(1, -1)
+        x = self._rand(self.space, shape)
+        _size = [] if self.shape == torch.Size([1]) else self.shape
+        return x.permute(*torch.arange(x.ndim - 1, -1, -1)).reshape([*shape, *_size])
 
     def _project(self, val: torch.Tensor) -> torch.Tensor:
         if val.dtype not in (torch.int, torch.long):
-            val = torch.round(val)
-        vals = self._split(val)
-        return torch.cat(
-            [
-                _val.clamp_(min=0, max=space.n - 1).unsqueeze(0)
-                for _val, space in zip(vals, self.space)
-            ],
-            dim=-1,
-        ).squeeze()
+            val = torch.round(val).type(self.dtype)
+        val = val.unsqueeze(0)
+        for permutation in itertools.product(*[range(d) for d in self.shape]):
+            val.unsqueeze(0)[[..., *permutation]] = val.unsqueeze(0)[
+                [..., *permutation]
+            ].clamp_(min=0, max=self.nvec[permutation] - 1)
+        return val.squeeze(0)
 
     def is_in(self, val: torch.Tensor) -> bool:
-        vals = self._split(val)
-        return self._size == len(vals) and all(
-            [
-                (0 <= _val).all() and (_val < space.n).all()
-                for _val, space in zip(vals, self.space)
-            ]
+        if val.ndim < 1:
+            val = val.unsqueeze(0)
+        val_have_wrong_dim = (
+            self.shape != torch.Size([1])
+            and val.shape[-len(self.shape) :] != self.shape
         )
+        if self.dtype != val.dtype or len(self.shape) > val.ndim or val_have_wrong_dim:
+            return False
+
+        for permutation in itertools.product(*[range(d) for d in self.shape]):
+            x = val.unsqueeze(0)[[..., *permutation]]
+            if not ((0 <= x).all() and (x < self.nvec[permutation]).all()):
+                return False
+        return True
 
     def to_onehot(self) -> MultOneHotDiscreteTensorSpec:
+        if len(self.shape) > 1:
+            raise RuntimeError(
+                f"DiscreteTensorSpec with shape != tensor.Size([1]) can't be converted OneHotDiscreteTensorSpec. Got "
+                f"shape={self.shape}."
```

Comment:
Can you add `"this could be accomplished via padding or nestedtensors but it is not implemented yet. If you would like to see that feature, please submit an issue of torchrl's github repo."`